<!DOCTYPE HTML>
<!--
   Fractal by HTML5 UP
   html5up.net | @ajlkn
   Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
   -->
<html>
   <title>Fractal by HTML5 UP</title>
   <meta charset="utf-8" />
   <meta name="viewport" content="width=device-width, initial-scale=1" />
   <meta name="description" content="">
   <meta name="author" content="">
   <link href="https://netdna.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css" rel="stylesheet">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
    <link rel="stylesheet" href="../fonts/font-mfizz.css">
    <link rel="stylesheet" href="../css/github-light.css">
    <link rel="stylesheet" href="../css/main.css">
   <!--[if lte IE 8]>
   <script src="../js/ie/html5shiv.js"></script>
   <![endif]-->   
   <!--[if lte IE 8]>
   <link rel="stylesheet" href="../css/ie8.css" />
   <![endif]-->
   <!--[if lte IE 9]>
   <link rel="stylesheet" href="../css/ie9.css" />
   <![endif]-->
   <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
      
      ga('create', 'UA-89369264-1', 'auto');
      ga('send', 'pageview');
      
   </script>
   </head>
   <body id="top"  data-spy="scroll" data-target="#toc" data-offset="100">
      <header id="header">
         <div class="jumbotron" style="background-color:transparent;">
         <img src="images/slamdunk_logo_light.png" width="400" style="margin-bottom:100px;">
         <h3>
            <b>Streamlining SLAMSeq analysis with ultra-high sensitivity</b>
         </h3>
         Slamdunk analyzes SLAMSeq data.
         <div class="row" style="margin-top:30px;">
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://www.github.com/t-neumann/slamdunk">
               <i class="fa fa-github fa-lg"></i> GitHub
               </a>
            </div>
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://pypi.python.org/pypi/slamdunk">
               <i class="icon-python fa-lg"></i> PyPI
               </a>
            </div>
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://hub.docker.com/r/tobneu/slamdunk-full">
               <i class="fa fa-docker"></i> Docker
               </a>
            </div>
            <div class="col-md-3">
               <a class="btn btn-block btn-lg btn-primary" href="https://slamdunk.readthedocs.io">
               <i class="fa fa-book fa-lg"></i> Docs
               </a>
            </div>
         </div>
      </header>
      <!-- Intro -->
      
      <div class="container docs-container">
      <div class="row">
        <div class="col-sm-9">
          <section id="intro" class="wrapper">
         <div class="inner">
            <span id="document-Introduction"></span><div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>Slamdunk maps and analyzes SLAM-Seq data.</p>
</div>
<span id="document-Installation"></span><div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>This section covers the installation of <em>slamdunk</em>. Alternatively one could also run <em>slamdunk</em> from out of the box <a class="reference internal" href="index.html#docker-label"><em>Docker</em></a> containers.</p>
<p>There are 2 different possibilities:</p>
<ol class="arabic simple">
<li>Installation from <a class="reference external" href="https://pypi.python.org/pypi">PyPI</a> using the <a class="reference internal" href="index.html#pip-label"><em>Python Package Index</em></a> <strong>(recommended)</strong></li>
<li>Installation from <a class="reference internal" href="index.html#source-label"><em>Source</em></a></li>
</ol>
<div class="section" id="requirements">
<h3>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h3>
<p>There are no major requirements for <em>slamdunk</em>. The python package will acquire all external dependencies by itself.</p>
<div class="section" id="r-runtime">
<h4>R runtime<a class="headerlink" href="#r-runtime" title="Permalink to this headline">¶</a></h4>
<p>Both <em>alleyoop</em> and <em>splash</em> utilize <a class="reference external" href="https://www.r-project.org/">R</a> for various calculations and diagnostic plots.
Therefore, a global installation of R is required to run those tools.</p>
<ul class="simple">
<li>Minimum required R version: <strong>R 3.2.2</strong></li>
<li>Required packages (from <a class="reference external" href="https://cran.r-project.org/">CRAN</a>):</li>
</ul>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">R packages</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><em>getopt</em></td>
</tr>
<tr class="row-odd"><td><em>ggplot2</em></td>
</tr>
<tr class="row-even"><td><em>gridExtra</em></td>
</tr>
<tr class="row-odd"><td><em>RColorBrewer</em></td>
</tr>
<tr class="row-even"><td><em>lattice</em></td>
</tr>
<tr class="row-odd"><td><em>matrixStats</em></td>
</tr>
<tr class="row-even"><td><em>dplyr</em></td>
</tr>
<tr class="row-odd"><td><em>tidyr</em></td>
</tr>
</tbody>
</table>
<p><strong>Note:</strong> All required R packages will be installed automatically from PyPI or upon the first analysis run of <em>alleyoop</em> or <em>splash</em> from source.</p>
</div>
</div>
<div class="section" id="python-package-index">
<span id="pip-label"></span><h3>Python Package Index<a class="headerlink" href="#python-package-index" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://pypi.python.org/pypi/pip">pip</a> is the recommended tool for installing Python packages. It allows a convenient and simple installation
of <em>slamdunk</em> from  the <a class="reference external" href="https://pypi.python.org/pypi">Python Package Index PyPI</a>.</p>
<div class="code bash highlight-python"><div class="highlight"><pre># Having root permissions

pip install --extra-index-url https://testpypi.python.org/pypi slamdunk

# Local user only

pip install --extra-index-url https://testpypi.python.org/pypi --user slamdunk
export PATH=$PATH:$HOME/.local/bin/
</pre></div>
</div>
<p><strong>Note:</strong> There is no official <em>slamdunk</em> release yet, so there will be a test version on <a class="reference external" href="https://testpypi.python.org/pypi">TestPyPI</a> not guaranteed to be in line with latest developments.
If you want to be absolutely sure you&#8217;re pulling the latest version, install <em>slamdunk</em> from <a class="reference internal" href="index.html#source-label"><em>Source</em></a>.</p>
</div>
<div class="section" id="source">
<span id="source-label"></span><h3>Source<a class="headerlink" href="#source" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li>Clone from <a class="reference external" href="https://github.com/t-neumann/slamdunk">Github</a>.</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>git clone https://github.com/t-neumann/slamdunk.git

cd slamdunk
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li>Install required python modules:</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre># Having root permissions
pip install -r requirements.txt

# Local user only
pip install --user -r requirements.txt
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li>Change to <cite>contrib</cite></li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>cd slamdunk/contrib
</pre></div>
</div>
<ol class="arabic simple" start="4">
<li>Install NGM by following the <cite>build-ngm.sh</cite> instructions.</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>./build-ngm.sh
</pre></div>
</div>
<ol class="arabic simple" start="5">
<li>Install VarScan2 following the <cite>build-varscan.sh</cite> instructions.</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>./build-varscan.sh
</pre></div>
</div>
<ol class="arabic simple" start="6">
<li>Install Samtools following the <cite>build-samtools.sh</cite> instructions.</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>./build-samtools.sh
</pre></div>
</div>
<ol class="arabic simple" start="7">
<li>Install RNASeqReadSimulator following the <cite>build-rnaseqreadsimulator.sh</cite> instructions.</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>./build-rnaseqreadsimulator.sh
</pre></div>
</div>
<ol class="arabic simple" start="8">
<li>Run <em>slamdunk</em> (optionally put it in your <em>$PATH</em>  to run it from anywhere).</li>
</ol>
<div class="code bash highlight-python"><div class="highlight"><pre>cd slamdunk/bin

# Run it from directory
./slamdunk --help

# Put it in your $PATH to run it from anywhere
export PATH=$(pwd):$PATH

slamdunk --help
</pre></div>
</div>
</div>
</div>
<span id="document-Quickstart"></span><div class="section" id="quickstart">
<h2>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h2>
<p>To analyze a given SLAMSeq dataset with <em>slamdunk</em> using the default parameters which worked well on datasets during development,
call the <em>all</em> dunk using the recommended minimum set of default parameters:</p>
<div class="code bash highlight-python"><div class="highlight"><pre>slamdunk all -r &lt;reference fasta&gt; -b &lt;bed file&gt; -o &lt;output directory&gt; -5 12 -n 100
             -t &lt;threads&gt; -m -rl &lt;maximum read length&gt; --skip-sam files [files ...]
</pre></div>
</div>
<p>The input data you need and parameters you need to set yourself are the following:</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="94%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><strong>-r</strong></td>
<td>The reference fasta file.</td>
</tr>
<tr class="row-odd"><td><strong>-b</strong></td>
<td>BED-file containing coordinates for 3&#8217; UTRs.</td>
</tr>
<tr class="row-even"><td><strong>-o</strong></td>
<td>The output directory where all output files of this dunk will be placed.</td>
</tr>
<tr class="row-odd"><td><strong>-t</strong></td>
<td>The number of threads to use for this dunk. NextGenMap runs multi-threaded, so it is recommended to use more threads than available samples (default: 1).</td>
</tr>
<tr class="row-even"><td><strong>files</strong></td>
<td>Samplesheet (see <a class="reference internal" href="index.html#sample-file"><em>Sample file format</em></a>) or a list of all sample BAM/FASTA(gz)/FASTQ(gz) files (wildcard * accepted).</td>
</tr>
</tbody>
</table>
<p>This will run the entire <em>slamdunk</em> analysis with the most relevant output files being:</p>
<ul class="simple">
<li>Folder <em>count</em>: Tab-separated <em>tcount</em> files containing the SLAMSeq statistics per UTR (see <a class="reference internal" href="index.html#tcount-file"><em>Tcount file format</em></a>)</li>
<li>Folder <em>filter</em>: BAM-files with the final mapped reads for visualization and further processing</li>
</ul>
<hr class="docutils" />
<p>If you want to have a very quick sanity check whether your desired conversion rates have been achieved, use this <em>alleyoop</em> command
to plot the UTR conversion rates in your sample:</p>
<div class="code bash highlight-python"><div class="highlight"><pre>alleyoop utrrates -o &lt;output directory&gt; -r &lt;reference fasta&gt; [-mq MQ] [-m] [-t &lt;threads&gt;]
                  -b &lt;bed file&gt; [-l MAXLENGTH] bam [bam ...]
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="95%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Parameter</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><strong>-o</strong></td>
<td>The output directory where the plots will be placed.</td>
</tr>
<tr class="row-odd"><td><strong>-r</strong></td>
<td>The reference fasta file.</td>
</tr>
<tr class="row-even"><td><strong>-mq</strong></td>
<td>Minimal base quality for SNPs.</td>
</tr>
<tr class="row-odd"><td><strong>-m</strong></td>
<td>Flag to activate the multiple T-&gt;C conversion stringency: Only T-&gt;C conversions in reads with more than 1 T-&gt;C conversion will be counted.</td>
</tr>
<tr class="row-even"><td><strong>-t</strong></td>
<td>The number of threads to use. All tools run single-threaded, so it is recommended to use as many threads as available samples.</td>
</tr>
<tr class="row-odd"><td><strong>-b</strong></td>
<td>BED-file containing coordinates for 3&#8217; UTRs.</td>
</tr>
<tr class="row-even"><td><strong>-l</strong></td>
<td>Maximum read length in all samples.</td>
</tr>
<tr class="row-odd"><td><strong>bam</strong></td>
<td>BAM file(s) containing the final filtered reads from the <em>filter</em> folder (wildcard * accepted).</td>
</tr>
</tbody>
</table>
<p>This will produce a plot like the following example where you can clearly see that individual conversions for a given starting base are balanced and unbiased,
except for the T-&gt;C conversions in this SLAMSeq sample with longer labelling times.</p>
</div>
<span id="document-Basics"></span><div class="section" id="basics">
<h2>Basics<a class="headerlink" href="#basics" title="Permalink to this headline">¶</a></h2>
<div class="section" id="conversion-rates">
<span id="conversionrates"></span><h3>Conversion rates<a class="headerlink" href="#conversion-rates" title="Permalink to this headline">¶</a></h3>
<p>The individual measured conversion rates within a 3&#8217; UTR are highly dependent on the T-content of a given UTR and the number of reads that cover those Ts.
The more reads containing Ts, the more accurate the overall estimation of the conversion rate of T&gt;Cs in this given 3&#8217; UTR will be.</p>
<p>We have adressed this by using a 3&#8217; UTR positions based view instead of a read based view: For each T-position in a given 3&#8217; UTR, we record
the number of reads covering this position and the number of reads with T&gt;C conversion at this position - the more reads, the more accurate the T&gt;C conversion rate estimation will be.</p>
<p>Both values are summed over all T-positions in the UTR - this way, less reliable T-positions with little reads will have lower impact on the overall T&gt;C conversion rate estimation than
T-positions with many reads.</p>
<a class="reference internal image-reference" href="_images/conversionrates.png"><img alt="_images/conversionrates.png" src="_images/conversionrates.png" style="width: 800px;" /></a>
<p><strong>Note:</strong> Since QuantSeq is a strand-specific assay, only sense reads will be considered for the final analysis!</p>
</div>
<div class="section" id="multimapper-reconciliation">
<h3>Multimapper reconciliation<a class="headerlink" href="#multimapper-reconciliation" title="Permalink to this headline">¶</a></h3>
<div class="section" id="definition-and-importance">
<h4>Definition and importance<a class="headerlink" href="#definition-and-importance" title="Permalink to this headline">¶</a></h4>
<p>Multimappers are reads which align to more than one position in the genome with equal best alignment scores.</p>
<p>Such reads make up a substantial part of the alignment and are generally ignored, but can provide valuable information if they can be unequivocally
reassigned to a given location, enhancing the overall signal and thus heavily influencing the overall accuracy.</p>
<p>However, there are also several caveats: For genes with isoforms or homologous genes, simply taking all alignments into account will heavily distort
the signal, since the reads will likely not account equally for all regions, but rather originate mainly from a single region which cannot be unambiguously assigned.</p>
</div>
<div class="section" id="reassignment-strategy">
<h4>Reassignment strategy<a class="headerlink" href="#reassignment-strategy" title="Permalink to this headline">¶</a></h4>
<p>We have settled for a very conservative multimapper reassignment strategy:</p>
<p>Since the QuantSeq technology specifically enriches for 3&#8217; UTRs, we only consider alignments to annotated 3&#8217; UTRs supplied to <em>slamdunk</em> as relevant.</p>
<p>Therefore, any multimappers with alignments to a single 3&#8217; UTR and non-3&#8217;UTR regions (i.e. not annotated in the supplied reference) will be unequivocally
assigned to the single 3&#8217;UTR. If there are multiple alignments to this single 3&#8217;UTR, one will be chosen at random.</p>
<p>For all other cases, were a read maps to several 3&#8217; UTRs, we are unable to reassign the read uniquely to a given 3&#8217;UTR and thus discard it from the analysis.</p>
<p>In short, the procedure is as follows, illustrated by the example below:</p>
<ol class="arabic simple">
<li>Create an <a class="reference external" href="https://pypi.python.org/pypi/intervaltree/2.0.4">Interval Tree</a> from the supplied 3&#8217; UTR annotation</li>
<li>Check the overlaps of all multimapping reads with 3&#8217; UTRs using the Interval Tree</li>
<li>Remove any reads with alignments to more than one single 3&#8217; UTR</li>
<li>Remove any additional alignments to non-UTR regions</li>
</ol>
<a class="reference internal image-reference" href="_images/multimappers.png"><img alt="_images/multimappers.png" src="_images/multimappers.png" style="width: 800px;" /></a>
</div>
</div>
         </div>
      </section>
        </div>
        <div class="col-sm-3" id="toc_column">
          <div id="toc" data-spy="affix" data-offset-top="254">
<ul class="nav nav-stacked">
<li><a href="#document-Introduction">Introduction</a></li>
<li><a href="#document-Installation">Installation</a></li>
<ul class="nav nav-stacked">
<li><a href="#requirements">Requiremenets</a></li>
<li><a href="#python-package-index">PyPI</a></li>
<li><a href="#source">Source</a></li>
</ul>
</ul>
<p class="backtotop"><a href="http://gitter.im/ewels/MultiQC">Discuss on Gitter</a></p>
            <p class="backtotop"><a href="#">Back to top</a></p>
          </div>
      </div>
    </div> <!-- /container -->
      
      
         <!-- Footer -->
      <footer id="footer">
         <div class="row">
         <div class="col-md-6">
            <ul class="icons">
               <a target="_blank" href="https://github.com/t-neumann"><i class="fa fa-github"></i> t-neumann</a>   |   
               <a target="_blank" href="https://github.com/philres"><i class="fa fa-github"></i> philres</a>
            </ul>
         </div>
         <div class="col-md-6">
            <p class="copyright">&copy; Tobias Neumann.<br/>Page template: <a href="http://html5up.net">HTML5 UP</a></p>
         </div>
      </footer>
      <!-- Scripts -->
    <script src="../js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="../js/jquery.min.js"></script>
    <script src="../js/jquery-1.11.3.min.js"></script>
      <script src="../js/jquery.scrolly.min.js"></script>
      <script src="../js/skel.min.js"></script>
      <script src="../js/util.js"></script>
      <!--[if lte IE 8]><script src="../js/ie/respond.min.js"></script><![endif]-->
      <script src="../js/main.js"></script>
      <script src="../js/docs.js"></script>
   </body>
</html>